### ISSUE TITLE
users-persistent test suite failure

### ISSUE THREAD

**Issue by bergmark** [agrafix/users#3]
```
test suite users-persistent-tests: RUNNING...
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))
Migrating: CREATE TABLE "login"("id" INTEGER PRIMARY KEY,"created_at" TIMESTAMP NOT NULL,"username" VARCHAR NOT NULL,"email" VARCHAR NOT NULL,"password" VARCHAR NOT NULL,"active" BOOLEAN NOT NULL,"more" BLOB NOT NULL,CONSTRAINT "unique_username" UNIQUE ("username"),CONSTRAINT "unique_email" UNIQUE ("email"))
Migrating: CREATE TABLE "login_token"("id" INTEGER PRIMARY KEY,"token" VARCHAR NOT NULL,"token_type" VARCHAR NOT NULL,"created_at" TIMESTAMP NOT NULL,"valid_until" TIMESTAMP NOT NULL,"owner" INTEGER NOT NULL REFERENCES "login",CONSTRAINT "unique_token" UNIQUE ("token"),CONSTRAINT "unique_typed_token" UNIQUE ("token","token_type"))

core user management
  should create valid users
  should not allow duplicates
  list and count should be correct
  updating and loading users should work
  deleting users should work
  reusing a deleted users name should work
initialisation
  calling initUserBackend multiple times should not result in errors
authentification
  auth as valid user with username should work
  auth as valid user with email should work
  auth with invalid credentials should fail FAILED [1]
  sessionless auth with valid userdata should work
  sessionless auth with invalid userdata should fail
  destroy session should really remove the session
  sessions should time out 1
  sessions should time out 2
password reset
  generates a valid token for a user
  a valid token should reset the password
  expired tokens should not do any harm
  invalid tokens should not do any harm
user activation
  activates a user with a valid activation token
  does not allow expired tokens to activate a user
  does not allow invalid tokens to activate a user

Failures:

  src/Web/Users/TestSpec.hs:121:
  1) authentification auth with invalid credentials should fail
       src/Web/Users/TestSpec.hs:123:
       expected: Nothing
        but got: Just (SessionId {unSessionId = "1c324a59-4d8b-4991-b6b9-d098e4b51a15"})

Randomized with seed 1295311266

Finished in 5.9110 seconds
22 examples, 1 failure
Test suite users-persistent-tests: FAIL
```

**Comment by bergmark** [comment]
Some more information, got this when building LTS Haskell 3.x, since there hasn't been a new release and it worked before, is it because a dependency changed in some unexpected way?

**Comment by agrafix** [comment]
Hm, that's interesting. I could not get it to fail with `lts-3.4`, trying `lts-3.5` now... What platform is this running?

**Comment by bergmark** [comment]
Sorry for a crappy explanation. I managed to reproduce it now!

I'm trying to build the next LTS release. I figured it was probably some dependency of this package that had a minor bump with a bug. I ran `cabal install --dry-run` on this project with lts-3.5 in a cabal.config to see which packages had updates, the only update that would be included in lts was bcrypt-0.0.7. It seems that bcrypt-0.0.7+memory-0.7 are incompatible. This problem doesn't appear on nightlies or when using the cabal solver since bcrypt-0.0.7+memory-0.10 will be picked instead.

So this doesn't block an lts release any longer, I'll exclude the new bcrypt there.

Ping @a1kmm
